File: 백준/Silver/1912. 연속합/연속합.cc
// 백준_실버2_DP_연속합

/*

*/

#include<iostream>

using namespace std;

int main() {

	int n;
	int* arr;
	int* dp;
	int sol;

	cin >> n;
	arr = new int[n + 1]{ 0, };
	dp = new int[n + 1]{ 0, };


	for (int i = 1; i <= n; i++) {
		cin >> arr[i];
	}

	dp[1] = arr[1];

	sol = dp[1];

	for (int i = 2; i <= n; i++) {
		if (dp[i - 1] + arr[i] > arr[i]) {
			dp[i] = dp[i - 1] + arr[i];
		}
		else {
			dp[i] = arr[i];
		}

		if (sol < dp[i]) {
			sol = dp[i];
		}
	}

	cout << sol;

	delete[] arr;
	delete[] dp;

	return 0;
}

File: 백준/Silver/1912. 연속합/연속합.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        String[] str = br.readLine().split(" ");
        int[] arr = new int[n+1];
        int[] dp = new int[n+1];
        int sol=0;

        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(str[i-1]);
        }

        sol = arr[1];
        for (int i=1; i<=n; i++) {
            if (dp[i-1] + arr[i] > arr[i]) {
                dp[i] = dp[i-1] + arr[i];
            }
            else {
                dp[i] = arr[i];
            }
            if (sol < dp[i]) {
                sol = dp[i];
            }
        }

        bw.write(sol+"");

        bw.flush();
        bw.close();

    }
}